server/lib/util.py

```diff
@@ -502,3 +505,112 @@ def gzip_compress_response(raw_content, is_json):
   if is_json:
     response.headers['Content-Type'] = 'application/json'
   return response
+
+
+def fetch_highest_coverage(parent_entity: str,
+                           child_type: str,
+                           variables: List[str],
+                           all_facets: bool,
+                           facet_ids: List[str] = []):
+  """
+  Fetches the latest available data with the best coverage for the given a
+  parent entity, child type, variables, and facets. If multiple variables are
+  passed in, selects dates with highest coverage independently for each
+  variable.
+
+  Response format:
+  {
+    "facets": {
+      <facet_id>: {<facet object>}
+    },
+    "data": {
+      <var_dcid>: {
+        <entity_dcid>: {
+          <observation point(s)>
+        }
+      }
+    }
+  }
+  """
+  MAX_DATES_TO_CHECK = 5
+  MAX_YEARS_TO_CHECK = 5
+  point_responses = []
+  series_dates_response = dc.get_series_dates(parent_entity, child_type,
+                                              variables)
+  facet_ids_set = set(facet_ids)
+  for observation_entity_counts_by_date in series_dates_response[
+      'datesByVariable']:
+    # Each observation_entity_counts_by_date contains the observation counts by date
+    best_coverage_date = _get_highest_coverage_date(
+        observation_entity_counts_by_date=observation_entity_counts_by_date,
+        facet_ids=facet_ids_set,
+        max_dates_to_check=MAX_DATES_TO_CHECK,
+        max_years_to_check=MAX_YEARS_TO_CHECK)
+    if not best_coverage_date:
+      continue
```

Comment:
can you check when regularly fetching observation points if caller expects all requested variables to be in the result?